[Part 1/n] Rename Dataset => Datastream in top level files and pydocs

ericl opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Context

As we are enabling streaming datasets by default in Ray 2.4, it's time to revisit the Datasets naming conundrum.

A few things we know for sure:
- "Ray Dataset" puts us in the same space as DataFrame / ETL libraries like Spark, which is a common confusion.
- Users are a little confused about "Ray Data" vs "Ray Dataset", however, this isn't a big deal--- more like a nit.

Goals:
- Avoid unnecessary churn and confusion
- Clearly position Ray Data as a specialized library for ML training ingest and batch inference
- Technical accuracy of the name (i.e., don't over-correct on positioning and come up with technically misleading component names).

## Proposal

1. Rename `Dataset` to `Datastream` (and secondary renamings like DatasetContext, DatasetIterator).
2. When a `Datastream` is cached or otherwise fully materialized in memory (e.g., created via from_arrow_refs), it is instead created as a subclass of Datastream: `MaterializedDatastream(Datastream)`.

This seems to solve the positioning challenges, while preserving the technical accuracy of the class names (i.e., it's never called a "stream" when there's no streaming involved).

## This PR

This PR starts by renaming the top-level classes and updating pydoc in the top-level files. In future PRs:
- [ ] Rename secondary APIs like DatasetContext, GroupedDataset, etc.
- [ ] Update internal references to Datastream type
- [ ] Rename dataset.py to datastream.py
- [ ] Update docs and example code

A few examples:
```
>>> ray.data.from_items([1,2,3,4,5])
MaterializedDatastream(num_blocks=5, num_rows=5, schema=<class 'int'>)

>>> ds = ray.data.range_tensor(10000)
Datastream(
   num_blocks=16,
   num_rows=10000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)

>>> ds.split(2)
[MaterializedDatastream(
   num_blocks=8,
   num_rows=5000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
), MaterializedDatastream(
   num_blocks=8,
   num_rows=5000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)]

>>> ds.cache()
MaterializedDatastream(
   num_blocks=16,
   num_rows=10000,
   schema={__value__: ArrowTensorType(shape=(1,), dtype=int64)}
)

```

c21 commented:
A little more research on other system, curious is the Datastream terminology here coming from Flink ([doc](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/overview/))?

Have you considered a name of `CachedDataStream` (this is same as [Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/operators/overview/#datastream-rarr-cacheddatastream)) or `DataStream(cached=True)`, compared to `InMemoryData`?

ericl commented:
I think a close analogy is Java 8 Streams: https://stackify.com/streams-guide-java-8/

In their API, Streams are a way to manipulate Collections. The Stream class represents a series of transformations on some base collection, similar to how in Ray Data the Datastream class represents a series of operations executed on some base input (either Datasource or InMemoryData). A Stream also be collected back to a Collection like a Datastream can be materialized into an InMemoryTable.

stephanie-wang commented:
> A little more research on other system, curious is the Datastream terminology here coming from Flink ([doc](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/overview/))?
> 
> Have you considered a name of `CachedDataStream` (this is same as [Flink](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/operators/overview/#datastream-rarr-cacheddatastream)) or `DataStream(cached=True)` (this is my first reaction), compared to `InMemoryData`?

I also prefer the `CachedDataStream` naming here over `InMemoryData`.

ericl commented:
Can you guys clarify why you like CachedDataStream? My mental model is this:

InMemoryData:
- from_items() -> InMemoryData   # This makes sense.
- cache() -> InMemoryData   # Not perfect but makes sense.
- split() -> List[InMemoryData]   # Not perfect but makes sense.

CachedDataStream:
- from_items() -> CachedDataStream   # This is odd.
- cache() -> CachedDataStream  # This makes sense.
- split() -> List[CachedDataStream]   # This is odd (implies the outputs are streaming, but they are not).

In other words, a more collection-oriented name seems to avoid sounding odd in all scenarios, even though it makes less sense for cache(). Btw, Flink's CachedDataStream is a lazily-cached stream, not a materialized data collection like we have here.

Another option is `MaterializedData[set]`, since "materialization" is a term I've seen users bring up organically, and it doesn't imply storage (e.g., "memory").

c21 commented:
btw a separate nit - can we name it `DataStream` instead of `Datastream`?

This would be same with rest of world (DataLoader, DataPipe, DataFrame, etc).

jianoaix commented:
I'd +1 to `MaterializedData`, because 1) once it's executed, it's not quite a "stream" anymore; 2) `MaterializedDataset` seems to bring back the `Dataset` concept/confusion. Maybe the only slight downside is that "materialize" doesn't match the "cache" verb in API name.

ericl commented:
@jianoaix I ended up calling it MaterializedDatastream in this PR, because it sounded better when writing the pydoc. MaterializedData / MaterializedDataset were harder to explain as return types.

c21 commented:
+1 to MaterializedDatastream, this sounds better than MaterializedData / MaterializedDataset.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -1589,44 +1598,44 @@ def train_test_split(
     @ConsumptionAPI(pattern="Args:")
-    def union(self, *other: List["Dataset[T]"]) -> "Dataset[T]":
-        """Combine this dataset with others of the same type.
+    def union(self, *other: List["Datastream[T]"]) -> "MaterializedDatastream[T]":
```
Can we keep the return value as `Datastream` instead of `MaterializedDatastream`? We will soon implement the union as a streaming physical operator (https://github.com/ray-project/ray/issues/32598) during Ray 2.5.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -2126,22 +2135,22 @@ def sort(
 
-    def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
-        """Zip this dataset with the elements of another.
+    def zip(self, other: "Datastream[U]") -> "MaterializedDatastream[(T, U)]":
```
Zip is lazy, so the return value should be `Datastream`.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -2172,14 +2181,14 @@ def zip(self, other: "Dataset[U]") -> "Dataset[(T, U)]":
     @ConsumptionAPI
-    def limit(self, limit: int) -> "Dataset[T]":
-        """Truncate the dataset to the first ``limit`` records.
+    def limit(self, limit: int) -> "MaterializedDatastream[T]":
```
Can we keep the return value as Datastream instead of MaterializedDatastream? We will soon implement the union as a streaming physical operator (https://github.com/ray-project/ray/issues/32761) during Ray 2.5.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -4055,52 +4062,53 @@ def __iter__(self):
         """
-        self._plan.execute(force_read=True)
-        return self
+        copy = Datastream.copy(self, _as=MaterializedDatastream)
+        copy._plan.execute(force_read=True)
+        return copy
```
I feel returning a new object would make the semantic a little confusing. `ds.cache()` is having side effect on `ds`, but `ds` itself is still a `Datastream` with `is_cached() == True`. Feeling overloaded that we have two objects

* `Datastream` with `is_cached() == True` (this is quite weird)
* `MaterializedDatastream`

```
>>> ds = ray.data.range(10)
>>> output_ds = ds.cache()
>>> output_ds
MaterializedDatastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
>>> ds.is_cached()
True
>>> ds
Datastream(num_blocks=10, num_rows=10, schema=<class 'int'>)
```

ericl replied:
I'll try to fix it so there isn't a side effect. That was not intended originally.

ericl replied:
Fixed by using a deep-copy.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -261,21 +262,28 @@ def __init__(
+    ) -> "Datastream[T]":
+        if not _as:
+            _as = Datastream
+        if _deep_copy:
+            return _as(ds._plan.copy(), ds._epoch, ds._lazy)
+        else:
+            return _as(ds._plan.deep_copy(), ds._epoch, ds._lazy)
```
shouldn't it be:

```
if _deep_copy:
    return _as(ds._plan.deep_copy(), ds._epoch, ds._lazy)
else:
    return _as(ds._plan.copy(), ds._epoch, ds._lazy)
```

ericl replied:
Fixed and added a test for it.

c21 commented on python/ray/data/dataset.py:
```diff
@@ -4047,52 +4056,53 @@ def __iter__(self):
-    @Deprecated(message="Use `Dataset.cache()` instead.")
-    def fully_executed(self) -> "Dataset[T]":
+    @Deprecated(message="Use `Datastream.cache()` instead.")
+    def fully_executed(self) -> "MaterializedDatastream[T]":
```
after this PR, `ds.fully_executed()/cache()` will have behavior change, that `ds.fully_executed()/cache()` will trigger execution each time of call. Before this PR, only first call of `ds.fully_executed()/cache()` triggers execution.

Do we want to document the behavior change somewhere in doc?

ericl replied:
I added a bit of explanation to the pydoc here, but I don't believe we need to further justify the previous behavior since that was never well defined / documented (in fact it seems like a bug).

c21 replied:
Ok, I was worried about people depending on the behavior. Cool, let's start with this.

ericl replied:
It seemed like this did affect one existing code path--- one test with DatasetPipeline with an optimization flag disabled. Since this path is not documented, I opted to remove it. We could also roll back the behavior (have the mutating behavior) as an alternative, and delay changing this until we remove DatasetPipeline. However, I think the likelihood someone is depending on this flag is low.

c21 replied:
It's fine with me to change the behavior here.
